File: litho-rendercore/src/main/java/com/facebook/rendercore/RenderTreeNode.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.rendercore;

import android.graphics.Rect;
import androidx.annotation.Nullable;
import com.facebook.proguard.annotations.DoNotStrip;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

@DoNotStrip
public class RenderTreeNode {

  private static final int DEFAULT_SIZE = 4;

  private final @Nullable RenderTreeNode mParent;
  private final RenderUnit mRenderUnit;
  private final @Nullable Object mLayoutData;
  private final Rect mBounds;
  private final int mAbsoluteX;
  private final int mAbsoluteY;
  private final @Nullable Rect mResolvedPadding;

  final int mPositionInParent;

  private List<RenderTreeNode> mChildren;

  public RenderTreeNode(
      final @Nullable RenderTreeNode parent,
      final RenderUnit renderUnit,
      final @Nullable Object layoutData,
      final Rect bounds,
      final @Nullable Rect resolvedPadding,
      final int positionInParent) {
    mParent = parent;
    mRenderUnit = renderUnit;
    mLayoutData = layoutData;
    mBounds = bounds;
    mAbsoluteX = parent != null ? parent.getAbsoluteX() + bounds.left : bounds.left;
    mAbsoluteY = parent != null ? parent.getAbsoluteY() + bounds.top : bounds.top;
    mResolvedPadding = resolvedPadding;
    mPositionInParent = positionInParent;
  }

  public void child(RenderTreeNode renderTreeNode) {
    if (mChildren == null) {
      mChildren = new ArrayList<>(DEFAULT_SIZE);
    }

    mChildren.add(renderTreeNode);
  }

  public Rect getBounds() {
    return mBounds;
  }

  public int getAbsoluteX() {
    return mAbsoluteX;
  }

  public int getAbsoluteY() {
    return mAbsoluteY;
  }

  /**
   * Sets the absolutes bounds of this render tree node in {@param outRect}; i.e. returns the bounds
   * of this render tree node within its {@link RootHost}.
   *
   * @param outRect the calculated absolute bounds.
   */
  public Rect getAbsoluteBounds(Rect outRect) {
    outRect.left = mAbsoluteX;
    outRect.top = mAbsoluteY;
    outRect.right = mAbsoluteX + mBounds.width();
    outRect.bottom = mAbsoluteY + mBounds.height();

    return outRect;
  }

  @Nullable
  public Rect getResolvedPadding() {
    return mResolvedPadding;
  }

  public RenderUnit getRenderUnit() {
    return mRenderUnit;
  }

  @Nullable
  public RenderTreeNode getParent() {
    return mParent;
  }

  public int getChildrenCount() {
    return mChildren != null ? mChildren.size() : 0;
  }

  public RenderTreeNode getChildAt(int idx) {
    return mChildren.get(idx);
  }

  public int getPositionInParent() {
    return mPositionInParent;
  }

  @Nullable
  public Object getLayoutData() {
    return mLayoutData;
  }

  public String generateDebugString(@Nullable RenderTree tree) {
    final long id = mRenderUnit.getId();
    final String contentType = mRenderUnit.getRenderContentType().toString();
    final int index = tree != null ? tree.getRenderTreeNodeIndex(mRenderUnit.getId()) : -1;
    final String bounds = mBounds.toShortString();
    final int childCount = getChildrenCount();
    final long parentId = mParent != null ? mParent.getRenderUnit().getId() : -1;

    return String.format(
        Locale.US,
        "Id=%d; contentType='%s'; indexInTree=%d; posInParent=%d; bounds=%s; absPosition=[%d, %d]; childCount=%d; parentId=%d;",
        id,
        contentType,
        index,
        mPositionInParent,
        bounds,
        mAbsoluteX,
        mAbsoluteY,
        childCount,
        parentId);
  }
}

File: litho-rendercore/src/main/java/com/facebook/rendercore/RenderTreeNode.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.rendercore

import android.graphics.Rect
import com.facebook.proguard.annotations.DoNotStrip
import java.util.Locale

@DoNotStrip
class RenderTreeNode(
    val parent: RenderTreeNode?,
    val renderUnit: RenderUnit<*>,
    val layoutData: Any?,
    val bounds: Rect,
    val resolvedPadding: Rect?,
    val positionInParent: Int
) {
  val absoluteX: Int = (parent?.absoluteX ?: 0) + bounds.left
  val absoluteY: Int = (parent?.absoluteY ?: 0) + bounds.top
  private val children: MutableList<RenderTreeNode> by
      lazy(LazyThreadSafetyMode.NONE) { ArrayList(DEFAULT_SIZE) }

  fun child(renderTreeNode: RenderTreeNode) {
    children.add(renderTreeNode)
  }

  /**
   * Sets the absolutes bounds of this render tree node in [outRect]; i.e. returns the bounds of
   * this render tree node within its [RootHost].
   *
   * @param outRect the calculated absolute bounds.
   */
  fun getAbsoluteBounds(outRect: Rect): Rect {
    outRect.left = absoluteX
    outRect.top = absoluteY
    outRect.right = absoluteX + bounds.width()
    outRect.bottom = absoluteY + bounds.height()
    return outRect
  }

  val childrenCount: Int
    get() = children.size

  fun getChildAt(idx: Int): RenderTreeNode {
    return children.get(idx)
  }

  fun generateDebugString(tree: RenderTree?): String {
    val id = renderUnit.id
    val contentType = renderUnit.description
    val index = tree?.getRenderTreeNodeIndex(renderUnit.id) ?: -1
    val bounds = bounds.toShortString()
    val childCount = childrenCount
    val parentId = parent?.renderUnit?.id ?: -1
    return String.format(
        Locale.US,
        "Id=%d; renderUnit='%s'; indexInTree=%d; posInParent=%d; bounds=%s; absPosition=[%d, %d]; childCount=%d; parentId=%d;",
        id,
        contentType,
        index,
        positionInParent,
        bounds,
        absoluteX,
        absoluteY,
        childCount,
        parentId)
  }

  companion object {
    private const val DEFAULT_SIZE = 4
  }
}
